### PR TITLE
Bug 1712279 -  Only expose pings and metrics methods that we want exposed

### PR DESCRIPTION
**Pull request by brizental**
Before:

image

After:

Screenshot 2022-02-21 at 16 01 11

### Pull Request checklist ###

- [x] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint` Runs _all_ linters
- [ ] ~**Tests**: This PR includes thorough tests or an explanation of why it does not~
- [ ] ~**Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one~
- [x] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work

### CODE REVIEW

**Comment by moz-glean** [comment]
# Build size report

Merging https://github.com/mozilla/glean.js/pull/1209 into [main](https://github.com/mozilla/glean.js) will:
* **Increase** the size of full **Web Extension** bundle build by `13%`.
* **Increase** the size of full **Website** bundle build by `13%`.
* **Increase** the size of full **Node.js** bundle build by `12%`.
* **Increase** the size of full **Qt/QML** bundle build by `8%`.

---

| | Current size | New size | Size increase |
|--:|:---:|:---:|:---:|
| **Web Extension** |
| core only | 52 KB | 52 KB | 📉 54 bytes |
| full bundle | 81 KB | 92 KB | 📉 11 KB |
| **Website** |
| core only | 53 KB | 53 KB | 📉 54 bytes |
| full bundle | 82 KB | 92 KB | 📉 11 KB |
| **Node.js** |
| core only | 52 KB | 52 KB | 📉 54 bytes |
| full bundle | 87 KB | 97 KB | 📉 11 KB |
| **Qt/QML** |
| core only | 68 KB | 74 KB | 📈 5.6 KB |
| full bundle | 68 KB | 74 KB | 📈 5.6 KB |

**Comment by moz-glean** [comment]
# Build size report

Merging https://github.com/mozilla/glean.js/pull/1209 into [main](https://github.com/mozilla/glean.js) will:
* **Increase** the size of full **Web Extension** bundle build by `12%`.
* **Increase** the size of full **Website** bundle build by `12%`.
* **Increase** the size of full **Node.js** bundle build by `11%`.
* **Increase** the size of full **Qt/QML** bundle build by `8%`.

---

| | Current size | New size | Size increase |
|--:|:---:|:---:|:---:|
| **Web Extension** |
| core only | 52 KB | 52 KB | 📉 418 bytes |
| full bundle | 81 KB | 91 KB | 📈 10 KB |
| **Website** |
| core only | 53 KB | 52 KB | 📉 418 bytes |
| full bundle | 82 KB | 92 KB | 📈 10 KB |
| **Node.js** |
| core only | 52 KB | 51 KB | 📉 418 bytes |
| full bundle | 87 KB | 97 KB | 📈 10 KB |
| **Qt/QML** |
| core only | 68 KB | 74 KB | 📈 5.5 KB |
| full bundle | 68 KB | 74 KB | 📈 5.5 KB |

File: glean/src/core/error/index.ts [hunk 1]
```diff
@@ -81,7 +81,7 @@ export default class ErrorManager {
    *        Defaults to the first value in `sendInPings`.
    * @returns The number of errors reported.
    */
-  async testGetNumRecordedErrors (
+  async testGetNumRecordedErrors(
```
Comment (Dexterp37) [inline glean/src/core/error/index.ts:84]: Ha! I wonder why the linter didn't complain :\

File: glean/src/core/metrics/types/counter.ts [hunk 2]
```diff
@@ -35,30 +35,24 @@ export class CounterMetric extends Metric<number, number> {
...
-   * This is absolutely not meant to be used outside of Glean itself.
-   * It may cause multiple issues because it cannot guarantee
-   * that the recording of the metric will happen in order with other Glean API calls.
-   *
-   * @param instance The metric instance to record to.
    * @param amount The amount we want to add.
    */
-  static async _private_addUndispatched(instance: CounterMetricType, amount?: number): Promise<void> {
-    if (!instance.shouldRecord(Context.uploadEnabled)) {
+  async addUndispatched(amount?: number): Promise<void> {
```
Comment (Dexterp37) [inline glean/src/core/metrics/types/counter.ts:54]: Technically speaking, this should not be used by external consumers. Does this make any difference here? I believe that, at the very least, we should leave the comment.

File: glean/src/core/metrics/types/datetime.ts [hunk 3]
```diff
@@ -173,19 +174,12 @@ class DatetimeMetricType extends MetricType {
...
-   * This is absolutely not meant to be used outside of Glean itself.
-   * It may cause multiple issues because it cannot guarantee
-   * that the recording of the metric will happen in order with other Glean API calls.
-   *
-   * @param instance The metric instance to record to.
    * @param value The date we want to set to.
    */
-  static async _private_setUndispatched(instance: DatetimeMetricType, value?: Date): Promise<void> {
-    if (!instance.shouldRecord(Context.uploadEnabled)) {
+  async setUndispatched(value?: Date): Promise<void> {
```
Comment (Dexterp37) [inline glean/src/core/metrics/types/datetime.ts:181]: This is not to be used outside of Glean.

File: glean/src/core/metrics/types/timespan.ts [hunk 4]
```diff
@@ -89,25 +91,18 @@ class TimespanMetricType extends MetricType {
...
-   * It may cause multiple issues because it cannot guarantee
-   * that the recording of the metric will happen in order with other Glean API calls.
+   * An implemention of `setRaw` that does not dispatch the recording task.
    *
-   * @param instance The metric instance to record to.
    * @param elapsed The elapsed time to record, in milliseconds.
    */
-  static async _private_setRawUndispatched(instance: TimespanMetricType, elapsed: number): Promise<void> {
-    if (!instance.shouldRecord(Context.uploadEnabled)) {
+  async setRawUndispatched(elapsed: number): Promise<void> {
```
Comment (Dexterp37) [inline glean/src/core/metrics/types/timespan.ts:98]: Should this be exposed?

File: glean/src/core/metrics/types/counter.ts (hunk 2 above)
Comment (brizental) [reply glean/src/core/metrics/types/counter.ts:54]: Before, these metric were exposed to users. Now users have literally no access to them unless they change Glean.js itself. Hm... But now that I write this, maybe we should leave the comment for future Glean.js developers. I'll bring them back.

File: glean/src/core/metrics/types/timespan.ts (hunk 4 above)
Comment (brizental) [reply glean/src/core/metrics/types/timespan.ts:98]: No. Note that I changed the exports of all metrics classes and this is inside the internal class, not exposed to users.

**Review by Dexterp37 — APPROVED** [review]

**Pull request merged**